File: java/profiler/src/main/java/io/runtimeconditions/profiler/extension/ArtifactValidator.java
package io.runtimeconditions.profiler.extension;

import io.runtimeconditions.profiler.manifest.ManifestModel;
import io.runtimeconditions.profiler.manifest.ManifestParser;
import io.runtimeconditions.profiler.manifest.YamlDocument;
import io.runtimeconditions.profiler.project.ArtifactDiscovery;
import io.runtimeconditions.profiler.project.RuntimeConditionsArtifact;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class ArtifactValidator {
    private static final String API_VERSION = "runtimeconditions.io/v1alpha1";
    private static final String EXTENSION_KIND = "RuntimeConditionsExtensionDefinition";
    private static final String BINDING_KIND = "RuntimeConditionsBinding";
    private static final String PACKAGE_KIND = "RuntimeConditionsPackage";
    private static final String LANGUAGE = "java";

    public List<ValidatedRuntimeConditionsArtifact> validate(List<RuntimeConditionsArtifact> artifacts) {
        List<ValidatedRuntimeConditionsArtifact> validated = new ArrayList<>();
        for (RuntimeConditionsArtifact artifact : artifacts) {
            validated.add(validateArtifact(artifact));
        }
        validateExtensionSet(validated);
        return validated;
    }

    private ValidatedRuntimeConditionsArtifact validateArtifact(RuntimeConditionsArtifact artifact) {
        List<RuntimeConditionsDiagnostic> diagnostics = new ArrayList<>();
        String manifestExtensionId = null;
        String extensionId = null;
        String extensionDefinitionUri = null;
        ExtensionDefinitionModel extensionDefinition = null;
        ManifestModel javaManifest = null;
        List<String> dependencies = List.of();
        YamlDocument manifest = null;

        try {
            manifest = YamlDocument.parse(readResource(artifact.manifestUri()));
        } catch (IOException | IllegalArgumentException e) {
            diagnostics.add(error("package-manifest", artifact.manifestUri(), "failed to read manifest: " + e.getMessage()));
        }

        if (manifest != null) {
            validateRequiredValue(manifest.scalar("apiVersion"), API_VERSION, "apiVersion", artifact.manifestUri(), diagnostics);
            String expectedKind = artifact.kind() == RuntimeConditionsArtifact.Kind.BINDING ? BINDING_KIND : PACKAGE_KIND;
            validateRequiredValue(manifest.scalar("kind"), expectedKind, "kind", artifact.manifestUri(), diagnostics);
            validateRequiredValue(manifest.scalar("metadata", "language"), LANGUAGE, "metadata.language", artifact.manifestUri(), diagnostics);
            if (!manifest.hasSection(LANGUAGE)) {
                diagnostics.add(error("package-language", artifact.manifestUri(), "java section is required"));
            } else {
                javaManifest = new ManifestParser().parse(manifest, artifact.manifestUri(), diagnostics);
            }

            if (artifact.kind() == RuntimeConditionsArtifact.Kind.BINDING) {
                manifestExtensionId = requireScalar(manifest, artifact.manifestUri(), diagnostics, "metadata", "extension");
                requireScalar(manifest, artifact.manifestUri(), diagnostics, "java", "package");
                extensionDefinitionUri = resolveExtensionDefinition(
                        artifact,
                        manifest.scalar("metadata", "extensionDefinition"),
                        diagnostics);
            } else {
                requireScalar(manifest, artifact.manifestUri(), diagnostics, "metadata", "package");
                manifestExtensionId = requireScalar(manifest, artifact.manifestUri(), diagnostics, "extension", "id");
                extensionDefinitionUri = resolveExtensionDefinition(
                        artifact,
                        manifest.scalar("extension", "definition"),
                        diagnostics);
            }
            validateExtensionId(manifestExtensionId, artifact.manifestUri(), diagnostics);
        }

        if (extensionDefinitionUri != null) {
            try {
                YamlDocument extension = YamlDocument.parse(readResource(extensionDefinitionUri));
                validateRequiredValue(extension.scalar("apiVersion"), API_VERSION, "apiVersion", extensionDefinitionUri, diagnostics);
                validateRequiredValue(extension.scalar("kind"), EXTENSION_KIND, "kind", extensionDefinitionUri, diagnostics);
                extensionId = requireScalar(extension, extensionDefinitionUri, diagnostics, "metadata", "id");
                validateExtensionId(extensionId, extensionDefinitionUri, diagnostics);
                extensionDefinition = ExtensionDefinitionModel.parse(extension, extensionDefinitionUri, diagnostics);
                dependencies = extension.stringList("spec", "dependencies");
            } catch (IOException | IllegalArgumentException e) {
                diagnostics.add(error("extension-definition", extensionDefinitionUri, "failed to read extension definition: " + e.getMessage()));
            }
        }

        if (manifestExtensionId != null && extensionId != null && !manifestExtensionId.equals(extensionId)) {
            diagnostics.add(error(
                    "extension-definition",
                    artifact.manifestUri(),
                    "manifest extension id " + manifestExtensionId + " does not match extension definition " + extensionId));
        }

        return new ValidatedRuntimeConditionsArtifact(
                artifact,
                manifestExtensionId,
                extensionId,
                extensionDefinitionUri,
                extensionDefinition,
                javaManifest,
                dependencies,
                diagnostics);
    }

    private void validateExtensionSet(List<ValidatedRuntimeConditionsArtifact> artifacts) {
        Map<String, String> definitionsById = new LinkedHashMap<>();
        Map<String, ExtensionDefinitionModel> definitionModelsById = new LinkedHashMap<>();
        for (ValidatedRuntimeConditionsArtifact artifact : artifacts) {
            if (artifact.extensionId() == null) {
                continue;
            }
            String previous = definitionsById.putIfAbsent(artifact.extensionId(), artifact.extensionDefinitionUri());
            if (previous != null && !previous.equals(artifact.extensionDefinitionUri())) {
                artifact.addDiagnostic(error(
                        "extension-definition",
                        artifact.extensionDefinitionUri(),
                        "duplicate extension id " + artifact.extensionId() + " already defined by " + previous));
            }
            if (artifact.extensionDefinition() != null) {
                definitionModelsById.putIfAbsent(artifact.extensionId(), artifact.extensionDefinition());
            }
        }
        for (ValidatedRuntimeConditionsArtifact artifact : artifacts) {
            for (String dependency : artifact.dependencies()) {
                if (!definitionsById.containsKey(dependency)) {
                    artifact.addDiagnostic(error(
                            "extension-dependency",
                            artifact.extensionDefinitionUri(),
                            "dependency " + dependency + " cannot be resolved from discovered Runtime Conditions artifacts"));
                }
            }
        }
        for (ValidatedRuntimeConditionsArtifact artifact : artifacts) {
            new ManifestVocabularyValidator().validate(artifact, definitionModelsById);
        }
    }

    private String resolveExtensionDefinition(
            RuntimeConditionsArtifact artifact,
            String overridePath,
            List<RuntimeConditionsDiagnostic> diagnostics) {
        if (overridePath != null && !overridePath.isBlank()) {
            try {
                return resolveRelative(artifact.manifestUri(), overridePath);
            } catch (IllegalArgumentException e) {
                diagnostics.add(error("extension-definition", artifact.manifestUri(), e.getMessage()));
                return null;
            }
        }
        if (artifact.extensionUri() == null) {
            diagnostics.add(error(
                    "extension-definition",
                    artifact.manifestUri(),
                    ArtifactDiscovery.EXTENSION_DEFINITION + " is required next to the manifest"));
            return null;
        }
        return artifact.extensionUri();
    }

    private String resolveRelative(String manifestUri, String relativePath) {
        if (manifestUri.startsWith("jar:")) {
            int separator = manifestUri.indexOf("!/");
            if (separator < 0) {
                throw new IllegalArgumentException("invalid JAR manifest URI: " + manifestUri);
            }
            String jarRoot = manifestUri.substring(0, separator + 2);
            String entry = manifestUri.substring(separator + 2);
            Path resolved = Path.of(entry).getParent().resolve(relativePath).normalize();
            if (resolved.startsWith("..")) {
                throw new IllegalArgumentException("extension definition override escapes the JAR artifact: " + relativePath);
            }
            return jarRoot + resolved.toString().replace('\\', '/');
        }
        URI uri = URI.create(manifestUri);
        if (!"file".equals(uri.getScheme())) {
            throw new IllegalArgumentException("unsupported manifest URI scheme for extension definition override: " + uri.getScheme());
        }
        return Path.of(uri).getParent().resolve(relativePath).normalize().toUri().toString();
    }

    private String readResource(String uriString) throws IOException {
        URI uri = URI.create(uriString);
        if ("file".equals(uri.getScheme())) {
            return java.nio.file.Files.readString(Path.of(uri));
        }
        if ("jar".equals(uri.getScheme())) {
            URL url = uri.toURL();
            try (InputStream input = url.openStream()) {
                return new String(input.readAllBytes(), StandardCharsets.UTF_8);
            }
        }
        throw new IOException("unsupported artifact URI scheme: " + uri.getScheme());
    }

    private String requireScalar(
            YamlDocument document,
            String source,
            List<RuntimeConditionsDiagnostic> diagnostics,
            String... path) {
        String value = document.scalar(path);
        if (value == null || value.isBlank()) {
            diagnostics.add(error("package-manifest", source, String.join(".", path) + " is required"));
            return null;
        }
        return value;
    }

    private void validateRequiredValue(
            String actual,
            String expected,
            String field,
            String source,
            List<RuntimeConditionsDiagnostic> diagnostics) {
        if (actual == null || actual.isBlank()) {
            diagnostics.add(error("package-manifest", source, field + " is required"));
        } else if (!expected.equals(actual)) {
            diagnostics.add(error("package-manifest", source, field + " must be " + expected));
        }
    }

    private void validateExtensionId(String id, String source, List<RuntimeConditionsDiagnostic> diagnostics) {
        if (id == null || id.isBlank()) {
            return;
        }
        try {
            URI uri = URI.create(id);
            if (!uri.isAbsolute() || uri.getHost() == null || (!"http".equals(uri.getScheme()) && !"https".equals(uri.getScheme()))) {
                diagnostics.add(error("extension-definition", source, "extension id must be an absolute HTTP or HTTPS URI"));
            }
        } catch (IllegalArgumentException e) {
            diagnostics.add(error("extension-definition", source, "extension id must be an absolute HTTP or HTTPS URI"));
        }
    }

    private RuntimeConditionsDiagnostic error(String code, String source, String message) {
        return new RuntimeConditionsDiagnostic(RuntimeConditionsDiagnostic.Severity.ERROR, code, source, message);
    }
}

File: python/profiler/runtimeconditions_profiler/extension/artifact_validator.py
from __future__ import annotations

from typing import Any, Optional
from urllib.parse import urlparse

from ..constants import API_VERSION, BINDING_KIND, EXTENSION_DEFINITION, EXTENSION_KIND, LANGUAGE, PACKAGE_KIND
from ..manifest.parser import ManifestParser, require_scalar
from ..models import ExtensionDefinition, RuntimeConditionsArtifact, ValidatedArtifact
from ..util import as_map, scalar, uri_to_path
from ..yamlio import Yaml
from .definition import dependency_cycle_errors, parse_extension_definition
from .manifest_validator import ManifestVocabularyValidator


class ArtifactValidator:
    def validate(self, artifacts: list[RuntimeConditionsArtifact]) -> list[ValidatedArtifact]:
        validated = [self._validate_one(artifact) for artifact in artifacts]
        self._validate_set(validated)
        return validated

    def _validate_one(self, artifact: RuntimeConditionsArtifact) -> ValidatedArtifact:
        item = ValidatedArtifact(artifact=artifact)
        manifest_doc: dict[str, Any] | None = None
        try:
            manifest_doc = Yaml.load(uri_to_path(artifact.manifest_uri))
        except Exception as exc:
            item.add("package-manifest", artifact.manifest_uri, f"failed to read manifest: {exc}")

        if manifest_doc is not None:
            expected_kind = BINDING_KIND if artifact.kind == "binding" else PACKAGE_KIND
            require_value(item, manifest_doc.get("apiVersion"), API_VERSION, artifact.manifest_uri, "apiVersion")
            require_value(item, manifest_doc.get("kind"), expected_kind, artifact.manifest_uri, "kind")
            metadata = as_map(manifest_doc.get("metadata"))
            require_value(item, metadata.get("language"), LANGUAGE, artifact.manifest_uri, "metadata.language")
            section = as_map(manifest_doc.get(LANGUAGE))
            if section is None:
                item.add("package-language", artifact.manifest_uri, "python section is required")
            else:
                item.manifest = ManifestParser().parse(section, artifact.manifest_uri, item)

            if artifact.kind == "binding":
                item.manifest_extension_id = require_scalar(item, metadata, artifact.manifest_uri, "metadata.extension")
                if section is not None:
                    require_scalar(item, section, artifact.manifest_uri, "python.package")
                override = scalar(metadata.get("extensionDefinition"))
            else:
                extension = as_map(manifest_doc.get("extension"))
                item.manifest_extension_id = require_scalar(item, extension, artifact.manifest_uri, "extension.id")
                override = scalar(extension.get("definition"))
            validate_extension_id(item, item.manifest_extension_id, artifact.manifest_uri)
            item.extension_definition_uri = resolve_extension_definition(artifact, override, item)

        if item.extension_definition_uri is not None:
            try:
                extension_doc = Yaml.load(uri_to_path(item.extension_definition_uri))
                require_value(item, extension_doc.get("apiVersion"), API_VERSION, item.extension_definition_uri, "apiVersion")
                require_value(item, extension_doc.get("kind"), EXTENSION_KIND, item.extension_definition_uri, "kind")
                metadata = as_map(extension_doc.get("metadata"))
                item.extension_id = require_scalar(item, metadata, item.extension_definition_uri, "metadata.id")
                validate_extension_id(item, item.extension_id, item.extension_definition_uri)
                if item.extension_id:
                    item.extension_definition = parse_extension_definition(
                        extension_doc,
                        item.extension_id,
                        item.extension_definition_uri,
                    )
                    item.dependencies = item.extension_definition.dependencies
            except Exception as exc:
                item.add("extension-definition", item.extension_definition_uri, f"failed to read extension definition: {exc}")

        if item.manifest_extension_id and item.extension_id and item.manifest_extension_id != item.extension_id:
            item.add(
                "extension-definition",
                artifact.manifest_uri,
                f"manifest extension id {item.manifest_extension_id} does not match extension definition {item.extension_id}",
            )
        return item

    def _validate_set(self, artifacts: list[ValidatedArtifact]) -> None:
        definitions_by_id: dict[str, str] = {}
        models_by_id: dict[str, ExtensionDefinition] = {}
        for artifact in artifacts:
            if not artifact.extension_id:
                continue
            previous = definitions_by_id.setdefault(artifact.extension_id, artifact.extension_definition_uri or "")
            if previous != (artifact.extension_definition_uri or ""):
                artifact.add(
                    "extension-definition",
                    artifact.extension_definition_uri or artifact.artifact.manifest_uri,
                    f"duplicate extension id {artifact.extension_id} already defined by {previous}",
                )
            if artifact.extension_definition is not None:
                models_by_id.setdefault(artifact.extension_id, artifact.extension_definition)

        for artifact in artifacts:
            for dependency in artifact.dependencies:
                if dependency not in definitions_by_id:
                    artifact.add(
                        "extension-dependency",
                        artifact.extension_definition_uri or artifact.artifact.manifest_uri,
                        f"dependency {dependency} cannot be resolved from discovered Runtime Conditions artifacts",
                    )

        cycle_errors = dependency_cycle_errors(models_by_id)
        for artifact in artifacts:
            for message in cycle_errors:
                artifact.add("extension-dependency", artifact.extension_definition_uri or artifact.artifact.manifest_uri, message)

        for artifact in artifacts:
            ManifestVocabularyValidator().validate(artifact, models_by_id)


def resolve_extension_definition(
    artifact: RuntimeConditionsArtifact,
    override: Optional[str],
    validated: ValidatedArtifact,
) -> Optional[str]:
    if override:
        path = (uri_to_path(artifact.manifest_uri).parent / override).resolve()
        return path.as_uri()
    if artifact.extension_uri is None:
        validated.add(
            "extension-definition",
            artifact.manifest_uri,
            f"{EXTENSION_DEFINITION} is required next to the manifest",
        )
        return None
    return artifact.extension_uri


def validate_extension_id(artifact: ValidatedArtifact, value: Optional[str], source: str) -> None:
    if not value:
        return
    parsed = urlparse(value)
    if parsed.scheme not in ("http", "https") or not parsed.netloc:
        artifact.add("extension-definition", source, "extension id must be an absolute HTTP or HTTPS URI")


def require_value(artifact: ValidatedArtifact, actual: Any, expected: str, source: str, field_name: str) -> None:
    parsed = scalar(actual)
    if not parsed:
        artifact.add("package-manifest", source, f"{field_name} is required")
    elif parsed != expected:
        artifact.add("package-manifest", source, f"{field_name} must be {expected}")

